xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -351,35 +351,83 @@ static final class StaticStrideScheduler {
     private final AtomicInteger sequence;
     private static final int K_MAX_WEIGHT = 0xFFFF;
 
+    // Assuming the mean of all known weights is M, StaticStrideScheduler will cap
+    // from above all known weights that are bigger than M*kMaxRatio
+    // to M*kMaxRatio and below all known weights to M*kMinRatio.
+    //
+    // This is done as a performance optimization by limiting the number of rounds for picks
+    // for edge cases where channels have large differences in subchannel weights.
+    // In this case, without these clips, it would potentially require the scheduler to
+    // frequently traverse through the entire subchannel list within the pick method.
+    //
+    // The current values of 10 and 0.1 were chosen without any experimenting. It should
+    // decrease the amount of sequences that the scheduler must traverse through in order
+    // to pick a high weight subchannel in such corner cases.
+    // But, it also makes WeightedRoundRobin to send slightly more requests to
+    // potentially very bad tasks (that would have near-zero weights) than zero.
+    // This is not necessarily a downside, though. Perhaps this is not a problem at
+    // all, and we can increase this value if needed to save CPU cycles.
+    //
+    // Note that this class treats weights that are exactly equal to zero as unknown
```

Comment:
Currently, negative/zero weights will be set to `mean` while weights that are close to zero (e.g. 0.000001), can be set to `mean*kMinRatio`. The EDFScheduler had a minimum weight of 0.0001, which we got rid of with the new scheduler. I'm getting confused on what's supposed to be "right" since we seem to be deviating from WRR with this weight clip.

But yes, I see what you mean about making this a TODO. What do you think the desired behavior here should be?

Reply:
It is normal good practice to not compare floats for equality directly but within a range of a small "epsilon", because rounding and accumulating error causes havoc. But we are comparing to literal zeros that are in the code; zero isn't the result of a computation. That makes it "work" to use equality checks, although not everyone would agree that it is "okay."

I was arguing this comment doesn't belong here. This is about the ratios which has nothing to do with "zero means use average." The code doesn't confuse the two cases. I'd be fine with this paragraph being deleted.